packages/firestore/test/unit/local/web_storage_metadata_channel.test.ts

```diff
@@ -0,0 +1,271 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * withOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as persistenceHelpers from './persistence_test_helpers';
+import {
+  InstanceStateSchema,
+  WebStorageMetadataChannel,
+  InstanceMetadataChannel
+} from '../../../src/local/instance_metadata_channel';
+import { BatchId, TargetId } from '../../../src/core/types';
+import { AutoId } from '../../../src/util/misc';
+import { expect } from 'chai';
+
+const GRACE_INTERVAL_MS = 100;
+
+describe('WebStorageMetadataChannel', () => {
+  if (!WebStorageMetadataChannel.isAvailable()) {
+    console.warn(
+      'No LocalStorage. Skipping LocalStorageNotificationChannelTests tests.'
+    );
+    return;
+  }
+
+  const localStorage = window.localStorage;
+
+  let notificationChannel: InstanceMetadataChannel;
+  let ownerId;
+
+  beforeEach(() => {
+    ownerId = AutoId.newId();
+  });
+
+  function assertInstanceState(
+    activeTargetIds: number[],
+    minMutationBatchId?: number,
+    maxMutationBatchId?: number
+  ): void {
+    const actual = JSON.parse(
+      localStorage.getItem(
+        `fs_instances_${persistenceHelpers.TEST_PERSISTENCE_PREFIX}_${ownerId}`
+      )
+    );
+
+    let expectedKeyCount = /* lastUpdateTime + activeTargetIds */ 2;
+
+    expect(actual.lastUpdateTime).to.be.a('number');
+    expect(actual.lastUpdateTime).to.be.greaterThan(
+      Date.now() - GRACE_INTERVAL_MS
+    );
+    expect(actual.lastUpdateTime).to.be.at.most(Date.now());
+
+    expect(actual.activeTargetIds).to.be.an('array');
+    expect(actual.activeTargetIds).to.be.deep.equal(activeTargetIds);
+
+    if (minMutationBatchId !== undefined) {
+      ++expectedKeyCount;
+      expect(actual.minMutationBatchId).to.be.equal(minMutationBatchId);
+    }
+
+    if (maxMutationBatchId !== undefined) {
+      ++expectedKeyCount;
+      expect(actual.maxMutationBatchId).to.be.equal(maxMutationBatchId);
+    }
+
+    expect(Object.keys(actual).length).to.be.equal(expectedKeyCount);
+  }
+
+  describe('persists mutation batches', () => {
+    beforeEach(() => {
+      return persistenceHelpers
+        .testLocalStorageNotificationChannel(ownerId, [], [])
+        .then(nc => {
+          notificationChannel = nc;
+        });
+    });
+
+    afterEach(() => {
+      notificationChannel.shutdown();
+    });
+
+    it('with empty batch', () => {
```

Comment:
"Persists mutation batches with empty batch" is a strange name.  Perhaps change this to 'when empty' or something?

Reply:
Sounds good